Enable Hardware-In-The-Loop simulation for JSBSim

Jaeyoung-Lim opened the pull request:
**Describe problem solved by this pull request**
Currently the `px4-jsbsim-bridge` can only be used to run Software-In-The-Loop(SITL) simulations. While this is useful for evaluating the software changes in firmware, it may not reveal issues that may appear only deployed on hardware. PX4 also supports [Hardware-In-The-Loop(HITL) simulation](https://dev.px4.io/master/en/simulation/hitl.html#hardware-in-the-loop-simulation-hitl). However, this was not exposed to the jsbsim-bridge

**Describe your solution**
This commit adds configuration options for running HITL simulation with the PX4 JSBSim bridge.

The device path / baudrate can be passed with the command line when running the binary directly with flag `-d` and `-b`.
e.g.
```
HEADLESS=1 ./jsbsim_bridge rascal -d /dev/ttyACM0 -s ~/src/Firmware/Tools/jsbsim_bridge/scene/LSZH.xml
```
or to set the baudrate
```
HEADLESS=1 ./jsbsim_bridge rascal -d /dev/ttyACM0 -b 921600 -s ~/src/Firmware/Tools/jsbsim_bridge/scene/LSZH.xml
```
The bridge is configured to be run with a HITL when the device path `-d` is provided. The baudrate is configured to be `921600` by default, but can be also configured through a `-b` flag.

**Test data / coverage**
[WIP]()

**Additional context**
- Fixes https://github.com/Auterion/px4-jsbsim-bridge/issues/27

bkueng commented on include/configuration_parser.h:
```diff
@@ -63,6 +60,9 @@ class ConfigurationParser {
   int getRealtimeFactor() { return _realtime_factor; }
+  bool getSerialEnabled() { return _serial_enabled; }
+  int getBaudrate() { return _baudrate; }
+  std::string getDevice() { return _device; }
```
```suggestion
  const std::string& getDevice() const { return _device; }
```

bkueng commented on src/configuration_parser.cpp:
```diff
@@ -54,28 +54,39 @@ bool ConfigurationParser::ParseEnvironmentVariables() {
+        break;
+      }
+      case 'b': {
+        _baudrate = std::stoi(std::string(optarg));
```
```suggestion
        _baudrate = atoi(optarg);
```

bkueng commented on src/jsbsim_bridge.cpp:
```diff
@@ -149,18 +149,35 @@ bool JSBSimBridge::SetFdmConfigs(ConfigurationParser &cfg) {
 }
 
-bool JSBSimBridge::SetMavlinkInterfaceConfigs(std::unique_ptr<MavlinkInterface> &interface, TiXmlHandle &config) {
+bool JSBSimBridge::SetMavlinkInterfaceConfigs(std::unique_ptr<MavlinkInterface> &interface, ConfigurationParser &cfg) {
```
```suggestion
bool JSBSimBridge::SetMavlinkInterfaceConfigs(std::unique_ptr<MavlinkInterface> &interface, const ConfigurationParser &cfg) {
```

bkueng commented on src/jsbsim_bridge.cpp:
```diff
@@ -149,18 +149,35 @@ bool JSBSimBridge::SetFdmConfigs(ConfigurationParser &cfg) {
 
-bool JSBSimBridge::SetMavlinkInterfaceConfigs(std::unique_ptr<MavlinkInterface> &interface, TiXmlHandle &config) {
+bool JSBSimBridge::SetMavlinkInterfaceConfigs(std::unique_ptr<MavlinkInterface> &interface, ConfigurationParser &cfg) {
+  TiXmlHandle config = *cfg.XmlHandle();
```
```suggestion
  const TiXmlHandle& config = *cfg.XmlHandle();
```

bkueng commented on src/jsbsim_bridge.cpp:
```diff
@@ -199,7 +216,12 @@ void JSBSimBridge::Run() {
 
   // Receive and handle actuator controls
-  _mavlink_interface->pollForMAVLinkMessages();
+  bool hil_mode_ = true;
```
always true (?)

Jaeyoung-Lim committed:
Enable HIL simulation for jsbsim

This commit adds configuration options for running HITL simulation with the PX4 JSBSim bridge

The device path / baudrate can be passed with the command line flag -d and -b. The bridge is configured to be run by HITL when the -d is provided. the baudrate is configured to be 921600 by default 


Enable HITL on mavlink interface

This PR enables HITL on the mavlink interface
Handle gcs udp port in config file

Jaeyoung-Lim committed:
Update src/configuration_parser.cpp

Co-authored-by: Beat Küng <beat-kueng@gmx.net>

fahadmaqbool16 commented:
@Jaeyoung-Lim . Hi I am interested in running HITL with JSBSim bridge, Can you please guide me with the steps to follow for this. any help will be highly appreciated.

Jaeyoung-Lim commented:
@fahadmaqbool16 The pull request already explains how to test this PR, what additional information do you need?

fahadmaqbool16 commented:
@Jaeyoung-Lim thanks for your reply, when I run 

> HEADLESS=1 ./jsbsim_bridge rascal -d /dev/ttyACM0 -s ~/src/Firmware/Tools/jsbsim_bridge/scene/LSZH.xml 

     in Tools directory I get 

> bash: ./Tools/jsbsim_bridge/: Is a directory

Nothing happens.

Jaeyoung-Lim commented:
@fahadmaqbool16 If you read the PR description, your command does not match what I have described

Also your log says that what you tried to execute is a directory, therefore it is expected that nothing happens

fahadmaqbool16 commented:
No, actually I was changing the directory to Tools to see if the command works, but it didn't. To clarify, I am 

> running this command HEADLESS=1 ./jsbsim_bridge rascal -d /dev/ttyACM0 -s ~/src/Firmware/Tools/jsbsim_bridge/scene/LSZH.xml

in the Tools directory, I get the error 

> ofp@ubuntu:~/PX4-Autopilot/Tools$ HEADLESS=1 ./jsbsim_bridge rascal -d /dev/ttyACM0 -s ~/src/Firmware/Tools/jsbsim_bridge/scene/LSZH.xml

> bash: ./jsbsim_bridge: Is a directory

if this command is not correct, can you please tell the correct command,

fahadmaqbool16 commented:
Hi, My bad, I was running the command in the Tools directory, now I am able to make the jsbsim bridge work by running that command (HEADLESS=1 ./jsbsim_bridge rascal -d /dev/ttyACM0 -s ~/src/Firmware/Tools/jsbsim_bridge/scene/LSZH.xml) from within the build directory (/PX4-Autopilot/build/px4_sitl_default/build_jsbsim_bridge). My pixhawk is detected on the serial port as shown in the image below.

![image](https://user-images.githubusercontent.com/94038272/191901668-a24485bf-fa60-4995-8dd0-c17a16ed89ab.png)

QGC is also able to connect with pixhawk but i am unable to arm the vehicle as the vehicle launch position is not showing in QGC.

![image](https://user-images.githubusercontent.com/94038272/191902013-16530dba-25aa-43e4-8d4d-f4592e39b989.png)

running dmesg in the MAVLINK console shows the follwing  output. 

![image](https://user-images.githubusercontent.com/94038272/191902179-f2e0be3c-b83f-4265-a7fa-48acda7f3690.png)

PS: I replaced the px4_jsbsim_bridge files from this repo. (https://github.com/Auterion/px4-jsbsim-bridge/tree/f88764b091a8f5bc5f98893f07c63b56682a1420)

Jaeyoung-Lim commented:
> PS: I replaced the px4_jsbsim_bridge files from this repo. (https://github.com/Auterion/px4-jsbsim-bridge/tree/f88764b091a8f5bc5f98893f07c63b56682a1420)

@fahadmaqbool16 You can't do that it only works with this specific branch

ParamDeshpande commented on src/jsbsim_bridge.cpp:
```diff
@@ -149,18 +149,35 @@ bool JSBSimBridge::SetFdmConfigs(ConfigurationParser &cfg) {
+    interface->SetSerialEnabled(cfg.getSerialEnabled());
+    interface->SetDevice(cfg.getDevice());
+    interface->SetBaudrate(cfg.getBaudrate());
+    interface->SetHILStateLevel(true);
```
interface->SetHILStateLevel(false); this fixes GPS Lock issue

cafrank commented on src/jsbsim_bridge.cpp:
std::shared_ptr<JSBSim::FGInitialCondition> initial_condition = _fdmexec->GetIC();
